### PR TITLE
Add spam memo exception for valid payers

### PR DESCRIPTION
**Pull request by vindard**
## Description

This PR builds on #415 and [this comment](https://github.com/GaloyMoney/galoy/pull/415#issuecomment-892184632) about valid transactions that could come in below the 'memo spam threshold'

### Approach

To solve this, @dolcalmi helped me to identify that the user with the `"funder"` role is usually the one sending these valid low-sat transactions.

Our approach was to add an additional step if a transaction is flagged below the memo threshold. We do a db call to lookup the username of the user that has the `role` property of `"funder`, and then compare this retrieved username to the `from` user's `username property gotten from the `transaction` object.

#### Implementation details

An additional db call is added asynchronously in the `getTransactions` method on the `UserWallet` object to retrieve the user with the `role: "funder"` property.

An earlier approach was tried where the additional lookup was included in the `memoSpamFilter` function itself only and only triggered (and then cached for that run of the method) if a txn amount was found _below the memo sats threshold_ to reduce the number of times we would hit the db.

This added more complexity and seemed to be mixing concerns to much, so I abandoned it given our upcoming work to clean things up and separate different concerns.

#### Potential future improvements

- To avoid this extra db call, we can also alternatively add an appropriate additional property (like `user role`) to transactions when they are created. **This would require a migration.**

- While implementing I got stuck with a circular dependency bug I introduced here that was fairly obfuscated for quite a while and didn't cause tests to fail until quite late in the run.
  
  **_Are there any linting or check we can add to catch circular deps at compile time?_**  (cc @samerbuna)
  
  _For e.g. I saw some folks use things like [madge](https://www.npmjs.com/package/madge) for this_

---
## Questions
1. Is the `username` property unique for users? (I assume that it is)

### CODE REVIEW

**Comment by dolcalmi** [comment]
> 1. Is the `username` property unique for users? (I assume that it is)
@vindard Yes, it is https://github.com/GaloyMoney/galoy/blob/main/src/services/mongoose/schema.ts#L161

File: src/core/user-wallet.ts [hunk 1]
```diff
@@ -102,22 +102,36 @@ export abstract class UserWallet {
...
         : null
 
-      const memoSpamFilter = (memoString) =>
-        memoString ? (isDebit || isValidCreditMemo ? memoString : null) : null
+      const memoSpamFilter = (memoString) => {
+        const creditBelowThreshold = isCredit && amount < MEMO_SHARING_SATS_THRESHOLD
+        const funderPayer =
+          item.username && item.username == lightningFundingUser?.username
+
+        const isCreditWithValidMemo = !creditBelowThreshold || funderPayer
```
Comment (dolcalmi) [inline src/core/user-wallet.ts:127]: I think maybe we can have a rule for boolean variables, For ex: must have `is` or `has` prefix so in that way are easier to read and identify the type

File: test/integration/02-user-wallet/02-send-lightning.spec.ts [hunk 2]
```diff
@@ -187,12 +203,30 @@ describe("UserWallet - Lightning Pay", () => {
...
+      memoFunderBelowThreshold,
+    )
+    expect(transaction4FunderBelow).toHaveProperty("username", userWallet0.user.username)
+    expect(transaction4FunderBelow).toHaveProperty(
+      "description",
+      memoFunderBelowThreshold,
+    )
+
     // check contacts being added
+    // TODO: decide if we want to exclude spam senders from contacts
```
Comment (dolcalmi) [inline test/integration/02-user-wallet/02-send-lightning.spec.ts:229]: this is a TODO but not in test, seems to be more a question for #dev channel in slack. Lets try to keep the code as clean as possible.

**Review by dolcalmi — CHANGES_REQUESTED** [review]
Just minor change(s). LGTM

**Comment by vindard** [comment]
This PR will be closed and the new functionality should be implemented based on this proposal: https://github.com/GaloyMoney/galoy/issues/363#issuecomment-900305025

**Pull request closed without merging**